packages/inferno-animation/src/animations.ts

```diff
@@ -25,66 +26,91 @@ function getAnimationClass(animationProp: AnimationClass | string | undefined |
+      // 2. Activate transition (after a reflow)
+      addClassName(dom, cls.active);
+      return;
+    case AnimationPhase.REGISTER_LISTENERS:
+      // 3. Set an animation listener, code at end
+      // Needs to be done after activating so timeout is calculated correctly
+      registerTransitionListener(
+        [dom],
+        function () {
+          // *** Cleanup ***
+          // 5. Remove the element
+          clearDimensions(dom);
+          removeClassName(dom, cls.active);
+          removeClassName(dom, cls.end);
 
-  // 4. Activate target state
-  requestAnimationFrame(() => {
-    setDimensions(dom, width, height);
-    removeClassName(dom, start);
-    addClassName(dom, end);
-  });
+          // 6. Call callback to allow stuff to happen
+          // Not currently used but this is where one could
+          // add a call to something like this.didAppearDone
+        }
+      );
+    case AnimationPhase.ACTIVATE_ANIMATION:
+      // 4. Activate target state (called async via requestAnimationFrame)
+      setDimensions(dom, dimensions.width, dimensions.height);
+      removeClassName(dom, cls.start);
+      addClassName(dom, cls.end);
+      return;
+  }
 }
 
 export function componentWillDisappear(dom: HTMLElement, props, callback: Function) {
-  const { start, end, active } = getAnimationClass(props.animation, '-leave');
-
-  // 1. Get dimensions and set animation start state
-  const { width, height } = getDimensions(dom);
-  setDimensions(dom, width, height);
-  addClassName(dom, start);
-
-  // 2. Activate transitions
-  addClassName(dom, active);
-
-  // 3. Set an animation listener, code at end
-  // Needs to be done after activating so timeout is calculated correctly
-  registerTransitionListener(
-    [dom],
-    function () {
-      // *** Cleanup not needed since node is removed ***
-      callback();
-    }
-  );
+  // Get dimensions and unpack class names
+  const cls = getAnimationClass(props.animation, '-leave');
+  const dimensions = getDimensions(dom);
+  queueAnimation((phase) => _willDisappear(phase, dom, callback, cls, dimensions));
+}
 
-  // 4. Activate target state
-  requestAnimationFrame(() => {
-    addClassName(dom, end);
-    removeClassName(dom, start);
-    clearDimensions(dom);
-  });
+function _willDisappear (phase: AnimationPhase, dom: HTMLElement, callback: Function, cls: AnimationClass, dimensions) {
+  switch (phase) {
+    case AnimationPhase.MEASURE:
+      // 1. Get dimensions and set animation start state
+      setDimensions(dom, dimensions.width, dimensions.height);
+      addClassName(dom, cls.start);
+      return;
+    case AnimationPhase.ACTIVATE_TRANSITIONS:
+      // 2. Activate transition (after a reflow)
+      addClassName(dom, cls.active);
+      return;
+    case AnimationPhase.REGISTER_LISTENERS:
+      // 3. Set an animation listener, code at end
+      // Needs to be done after activating so timeout is calculated correctly
+      registerTransitionListener(
```

Comment:
same comment about moving this function call outside because it creates function context. 

Or if the function scope is not necessary we could change this:
```js
registerTransitionListener([dom], callback);
```

Reply:
WILL INVESTIGATE